### ISSUE TITLE
Launch as http://localhost instead of http://127.0.0.1?

### ISSUE THREAD

**Issue by mmdolbow** [microsoft/vscode-livepreview#631]
Recently switched to this extension instead of [preview on live server](https://marketplace.visualstudio.com/items?itemName=yuichinukiyama.vscode-preview-server), and like it overall. But there's one thing either missing - or I can't figure out how to configure it: launching in the preview/browser as  http://localhost instead of http://127.0.0.1. This would allow better support for pulling in cross-origin resources like FontAwesome, which allows development on localhost but not on a specific IP address.

I'm thinking it's possible I just am missing a configuration option because [this issue](https://github.com/microsoft/vscode-livepreview/issues/145) mentions localhost several times. Is there maybe a more general VS Code or workspace setting I can use to ensure that I'm launching on localhost instead of the specific IP?

**Comment by andreamah** [comment]
Currently, there is no setting for opening the preview on localhost. Are you using an endpoint that only allows access from the localhost domain name? And if so, are there no workarounds with this? 

If you're hosting locally, there are options in `livePreview.httpHeaders` to modify headers (and potentially the CORS headers) if that helps in your case for consumption.

**Comment by mmdolbow** [comment]
> Currently, there is no setting for opening the preview on localhost. 

Bummer. Well at least now I know not to look for something that isn't there!

>Are you using an endpoint that only allows access from the localhost domain name? And if so, are there no workarounds with this?

It's not that the resource (FontAwesome in particular) only allows access from localhost, but that it only allows access any domain specified. Attempting to specify an IP address doesn't work. Since there's no direct reference (that I can find) to that restriction in their docs, I'm attaching a screen shot. I'm not aware of any workarounds.
![font-awesome-error-ip](https://github.com/microsoft/vscode-livepreview/assets/8397589/f562697e-d74e-4373-a72b-e33d0962038c)


> If you're hosting locally, there are options in `livePreview.httpHeaders` to modify headers (and potentially the CORS headers) if that helps in your case for consumption.

Thanks - I'll try that to see if I can get it to work.

**Comment by mmdolbow** [comment]
> If you're hosting locally, there are options in `livePreview.httpHeaders` to modify headers (and potentially the CORS headers) if that helps in your case for consumption.

Wait, my understanding is that these options, as you have in the help for the settings, modify "The extra HTTP headers that should be set in the server's HTTP responses." So these headers impact what the browser sees when requesting resources from the **local host**, ie http://127.0.0.1:3000.

What I need to manipulate is what ``Referer`` is specified when the browser then requests an outside resource (ie. AJAX or script source). In other words, **Request Headers**, not **Response Headers**. I'm attaching another screen grab from Firefox's console.
![font-awesome-request-headers](https://github.com/microsoft/vscode-livepreview/assets/8397589/f2213f5d-4e7f-4d96-a926-7b6503bba347)

If the extension merely allowed launch on localhost in addition to the specific IP, I wouldn't need to manipulate those headers, they would just be there. I've been using [Live Server](https://github.com/ritwickdey/vscode-live-server-plus-plus) instead as a result of this requirement.

Totally understand if you don't want to match that requirement! As I mentioned above, more importantly I wanted to make sure I wasn't spending time looking for an option that didn't exist.

**Comment by andreamah** [comment]
I can make this a feature for the backlog to support localhost for launching, but I cannot guarantee that I can get to it soon. I can see the setting enabling launching localhost if `livePreview.hostIP` is set to `127.0.0.1`

**Comment by D2G-TimP** [comment]
I stumbled on this issue while working with Live Preview and am running an integration with a 3rd party payment provider sandbox that allows http to be used only with localhost. It does not recognize the loopback address as localhost. 

I've been working around this, but being able to have it launch using localhost would be an awesome feature for this extension. 

Not trying to expedite, just trying to add further justification to why this functionality would be needed. 
Thank you for helping to make VS Code even better.

**Comment by nickjuntilla** [comment]
This is also necessary for any javascript service workers.

**Comment by max38tech** [comment]
Wow, a year and still no localhost. A lot of endpoints don't allow 127.0.0.1 but do allow localhost. What's it going to take? MS doesn't care? If you give me the source for this I'll fix it today. Not to sound ungrateful, I love the live preview thing, but not being able to actually TEST the page functions means I simply can't use it. So what's that worth? At least I'm offering to fix it.

**Comment by mmdolbow** [comment]
> Wow, a year and still no localhost. A lot of endpoints don't allow 127.0.0.1 but do allow localhost. 

I've just been using [Live Server](https://marketplace.visualstudio.com/items/?itemName=ritwickdey.LiveServer) instead. I wish every extension I tried had everything I need, but sometimes I just gotta switch!

**Comment by max38tech** [comment]
Yeah, that's what I figure I'll do. Probably not everything I do will have a problem with the local IP. 
Cheers

**Comment by caiohamamura** [comment]
As "funny" as it may sound I found out the issue. The issue is that in connectionManager.ts the code is expecting the host to be an IPv4 address. I don't know exactly why the developers wanted to enforce that the only valid host should be IP addresses and not custom domains or at least localhost.

Just changed this:

https://github.com/microsoft/vscode-livepreview/blob/4534410e2f2bd26526aae444773647cad1ed1c8f/src/connectionInfo/connectionManager.ts#L150

to:

https://github.com/caiohamamura/vscode-livepreview/blob/756d4ed5fcc32748f22828fbc2c38f7df2a4a9b6/src/connectionInfo/connectionManager.ts#L150

And it works.

**Comment by caiohamamura** [comment]
Can anyone from the community of the original contributors (@andreamah ?) help me to understand the strict validation of the host? Could we loosen that to allow IPv4 or at least 'localhost' as valid? 

From what I've tested here that is the only thing that is avoiding us set 
```
{
    "livePreview.hostIP": "localhost"
}
```

Everything else is working fine here. If we do not allow to set up "localhost", even if we open `localhost` instead of 127.0.0.1 in the browser then the `ws` connection will fail because the domain is not the same as setup, so the live reload will break. 

This is a major issue in many development scenarios, because there are lots of libraries and APIs that do allow CORS, but not from IP addresses, such as [fontawesome](https://fontawesome.com/), [imgur](imgur.com), ...
